### ISSUE TITLE
shinyproxy whitescreen after page reload

### ISSUE THREAD

**Issue by alex-d13** [openanalytics/shinyproxy#306]
Hi,

I am having issues with shinyproxy:
When my container is started and running and I reload the browser page, I get a whitescreen (in the network console: error 503; service not availabe). Then after 6s (I reduced the heartbeat to 6000), after the container is released due to inactivity and I reload the page again, a new container is started.
Is there a way to connect to the original container, without having to wait for 6s until the container was released and a new one is started?

Here is my application.yml

```
proxy:
  port: 8080
  landing-page: /app/namco
  authentication: none
  admin-groups: admins
  container-log-path: ./container-logs
  container-backend: docker
  container-wait-time: 10000
  heartbeat-rate: 1000 # the user's browser will send a heartbeat call every heartbeat-rate milliseconds (10s as default)
  heartbeat-timeout: 6000 # if the server does not receive a heartbeat for heartbeat-timeout milliseconds, the relevant proxy will be released (60s as default)
  hide-navbar: true
  users:
  - name: jack
    password: password
    groups: admins
  docker:
    internal-networking: true
  specs:
  - id: namco
    display-name: Namco
    description: Webapp for analysis of microbiome data
    container-cmd: ["R", "-e", "shiny::runApp('/srv/shiny-server/', host = '0.0.0.0', port = 3838)"]
    #container-cmd: ["/usr/bin/shiny-server.sh"]
    container-network: namco_sp-net
    container-image: namco

logging:
  file:
    name: ./shinyproxy.log

server:
  useForwardHeaders: true
```

and my nginx.conf

```
events {}
http {

    upstream namco {
    server shinyproxy:8080;
    }

    server {

    listen 80;
    server_name  localhost;

    underscores_in_headers on;


    location / {
        proxy_pass          http://namco;

        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_read_timeout 600s;

        proxy_redirect    off;
        proxy_set_header  Host             $http_host;
        proxy_set_header  X-Real-IP        $remote_addr;
        proxy_set_header  X-Forwarded-For  $proxy_add_x_forwarded_for;
        proxy_set_header  X-Forwarded-Protocol $scheme;
    }
    }


}

```

Does anyone have an idea what is going on here?

Cheers,
Alex

**Comment by LEDfan** [comment]
Hi @alex-d13 

Reloading the page should work. Unfortunately, I don't know what is causing your problem. Could you check the logs of nginx and ShinyProxy when the 503 error happens?

**Comment by alex-d13** [comment]
Nothing really going on in the shinyproxy console:
```
2021-07-08 06:59:52.711  INFO 1 --- [XNIO-1 task-3] c.s.docker.client.DefaultDockerClient    : Starting container with Id: 2c696bf95cf631b943d631d77c22023c963fd415a274a525c55e193c945f8413
2021-07-08 06:59:57.675  INFO 1 --- [XNIO-1 task-3] e.o.containerproxy.service.ProxyService  : Proxy activated [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 8fcbadab-4869-4005-bb9d-1
37accfdf274]
2021-07-08 09:04:04.565  INFO 1 --- [InactiveProxyKiller] e.o.c.service.HeartbeatService           : Releasing inactive proxy [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 8fcbadab-4
869-4005-bb9d-137accfdf274] [silence: 7849ms]
2021-07-08 09:04:04.592  INFO 1 --- [pool-2-thread-1] e.o.containerproxy.service.ProxyService  : Proxy released [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 8fcbadab-4869-4005-bb9d-
137accfdf274]
2021-07-08 09:05:42.837  INFO 1 --- [XNIO-1 task-3] c.s.docker.client.DefaultDockerClient    : Starting container with Id: c3a75439741fe89a5718119e062dafef16f4adc04793c7bc8ff468d25444d622
2021-07-08 09:05:47.626  INFO 1 --- [XNIO-1 task-3] e.o.containerproxy.service.ProxyService  : Proxy activated [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: ad09f3c5-9677-42cc-873d-f
48e6f78f5a6]
2021-07-08 09:05:56.582  INFO 1 --- [InactiveProxyKiller] e.o.c.service.HeartbeatService           : Releasing inactive proxy [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: ad09f3c5-9
677-42cc-873d-f48e6f78f5a6] [silence: 7354ms]
2021-07-08 09:05:57.085  INFO 1 --- [pool-2-thread-1] e.o.containerproxy.service.ProxyService  : Proxy released [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: ad09f3c5-9677-42cc-873d-
f48e6f78f5a6]
2021-07-08 09:07:21.778  INFO 1 --- [XNIO-1 task-1] c.s.docker.client.DefaultDockerClient    : Starting container with Id: 1cb146463ebc56782238f90dde2cd94bac028b49e7b541e40a90d5e0fb83794c
2021-07-08 09:07:26.445  INFO 1 --- [XNIO-1 task-1] e.o.containerproxy.service.ProxyService  : Proxy activated [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 040b7960-8b27-4b4b-b9b3-3
35083453f1c]
2021-07-08 09:08:08.605  INFO 1 --- [InactiveProxyKiller] e.o.c.service.HeartbeatService           : Releasing inactive proxy [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 040b7960-8
b27-4b4b-b9b3-335083453f1c] [silence: 6207ms]
2021-07-08 09:08:08.623  INFO 1 --- [pool-2-thread-1] e.o.containerproxy.service.ProxyService  : Proxy released [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 040b7960-8b27-4b4b-b9b3-
335083453f1c]
2021-07-08 09:08:14.497  INFO 1 --- [XNIO-1 task-1] c.s.docker.client.DefaultDockerClient    : Starting container with Id: f2e6b0926085c97930abb1364d1249c5ff30df9029094ae6a65a9e2f3be876ee
2021-07-08 09:08:19.171  INFO 1 --- [XNIO-1 task-1] e.o.containerproxy.service.ProxyService  : Proxy activated [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: b8ce09e1-63f8-4ea9-9b83-a
9a11a6b6e68]
2021-07-08 09:08:28.610  INFO 1 --- [InactiveProxyKiller] e.o.c.service.HeartbeatService           : Releasing inactive proxy [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: b8ce09e1-6
3f8-4ea9-9b83-a9a11a6b6e68] [silence: 7735ms]
2021-07-08 09:08:29.185  INFO 1 --- [pool-2-thread-1] e.o.containerproxy.service.ProxyService  : Proxy released [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: b8ce09e1-63f8-4ea9-9b83-
a9a11a6b6e68]
2021-07-08 09:08:29.212  INFO 1 --- [XNIO-1 task-5] c.s.docker.client.DefaultDockerClient    : Starting container with Id: 44a908816bf1e65504038b5e09c7b80190449b460f94c594b2bd006dd3adff56
2021-07-08 09:08:33.922  INFO 1 --- [XNIO-1 task-5] e.o.containerproxy.service.ProxyService  : Proxy activated [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 6937058d-b135-4ee6-bfdc-6
0e2c6bf5556]
2021-07-08 09:10:36.632  INFO 1 --- [InactiveProxyKiller] e.o.c.service.HeartbeatService           : Releasing inactive proxy [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 6937058d-b
135-4ee6-bfdc-60e2c6bf5556] [silence: 6059ms]
2021-07-08 09:10:36.651  INFO 1 --- [pool-2-thread-1] e.o.containerproxy.service.ProxyService  : Proxy released [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 6937058d-b135-4ee6-bfdc-
60e2c6bf5556]
2021-07-08 09:10:41.091  INFO 1 --- [XNIO-1 task-2] c.s.docker.client.DefaultDockerClient    : Starting container with Id: df3598e80437188510c52325f0a3438639be5b8d6e59a2ac1723bba9b00780f5
2021-07-08 09:10:45.787  INFO 1 --- [XNIO-1 task-2] e.o.containerproxy.service.ProxyService  : Proxy activated [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 66f4719b-169b-42a8-ab96-4
aa151c3d6a1]
2021-07-08 09:11:36.644  INFO 1 --- [InactiveProxyKiller] e.o.c.service.HeartbeatService           : Releasing inactive proxy [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 66f4719b-1
69b-42a8-ab96-4aa151c3d6a1] [silence: 7415ms]
2021-07-08 09:11:36.662  INFO 1 --- [pool-2-thread-1] e.o.containerproxy.service.ProxyService  : Proxy released [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 66f4719b-169b-42a8-ab96-
4aa151c3d6a1]
2021-07-08 09:12:08.550  INFO 1 --- [XNIO-1 task-3] c.s.docker.client.DefaultDockerClient    : Starting container with Id: 0e8af1b41d6dae84ebcdb29e01154bfe4c707df0df1cd63eff45fc437c827968
2021-07-08 09:12:13.714  INFO 1 --- [XNIO-1 task-3] e.o.containerproxy.service.ProxyService  : Proxy activated [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 5a1dc66b-1660-46f5-a9d8-3
13a91a9de51]
2021-07-08 09:12:44.656  INFO 1 --- [InactiveProxyKiller] e.o.c.service.HeartbeatService           : Releasing inactive proxy [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 5a1dc66b-1
660-46f5-a9d8-313a91a9de51] [silence: 6897ms]
2021-07-08 09:12:44.674  INFO 1 --- [pool-2-thread-1] e.o.containerproxy.service.ProxyService  : Proxy released [user: CAL0T54DQSI4GPjl5AHx7PQJ3iFw6vVKt9pmwOIM] [spec: namco] [id: 5a1dc66b-1660-46f5-a9d8-
313a91a9de51]
```
I also found this error in the browser network console, might have something to do with it?

```
The resource from “http://localhost:3848/app_direct/namco/” was not rendered due to an unknown, incorrect or missing MIME type (X-Content-Type-Options: nosniff).
```

**Comment by LEDfan** [comment]
Hi

Can you provide your `application.yml` file? (with the sensitive parts removed). Are you using `none` as authentication type (i.e. no-login)? And finally: are you running a proxy or load-balancer in front of ShinyProxy? Thanks!

**Comment by alex-d13** [comment]
Hi,

the application.yml is provided in my first post, and yes, I am using `none` as authentication.
I am using nginx as a proxy in front, the config file is also added in my first post.

Alex

**Comment by LEDfan** [comment]
Hi

Which version of ShinyProxy are you using? If you are using 2.4.0 or later, please change the configuration for the forward headers: https://shinyproxy.io/documentation/security/#forward-headers . 
If that doesn't solve the issue, can you try opening the `http://localhost:3848/app_direct/namco` directly and see if that works? Unfortunately I don't exactly know what is causing the issue, so we will have to debug it together.

**Comment by alex-d13** [comment]
Hi

I tried out both suggestions (I am using version 2.4.0 and was using `foreward-headers: true`, so I changed that).
Do you think it would help if I send you the Docker image I am using? I am working togther with other people which did not have this issue for other shiny applications but we basically are using the same shinyproxy setup, so maybe the issue lies within the shiny image?

**Comment by LEDfan** [comment]
Hi

Yes you can pass me the docker image. Nevertheless, can you you explain/show what happened when you opened `http://localhost:3848/app_direct/namco`? Can you maybe take a screenshot of the network tab of the browser developer tools (e.g. F12 -> click Network tab and reload the page). Thanks!

**Comment by Tylermarques** [comment]
Hey @LEDfan,

Any chance you ever made any progress on this issue? Experiencing something similar in in 2.6.1. Some users periodically experience a white screen when attempting to login, which appears to be caused by a 503 error to the https://{URL}/app_direct_i/{app_name}/_/ url. I'm not sure what is causing this. 

The app continues to send heartbeats, but the service appears to be unavailable. Does ShinyProxy check to ensure that the service is still available? It doesn't look like it from what I can see in the heartbeat service but I am by no means a Java dev.

**Comment by LEDfan** [comment]
Hi @Tylermarques 

No, unfortunately I was not able to reproduce this issue. You are correct that ShinyProxy does not actively check whether an app is still alive. The heartbeat mechanism is purely to detect whether the user is still using an app. 

Could you share the application you are using? This could help in debugging the issue.

**Comment by Tylermarques** [comment]
I unfortunately can't share the application I have, but I am fairly certain that the issue is due to Shiny containers closing unexpectedly while ShinyProxy assumes they are still running. When a user who has an active session attempts to visit the application, ShinyProxy does not check if that container is still active and that results in a 503 when attempting to connect to the websocket for a Shiny container that is no longer running. 

In my specific case, my shiny containers were getting OOM killed when a user loaded a particularly large data set. The shiny front end would grey out, which I believe is the state when the websocket is no longer responded. Users would then refresh and ShinyProxy then attempted to load the user back into their session (which hadn't timed out yet) and the users were instead presented with a blank white screen. 

To me this seems more as a feature request than a bug. It would be nice for ShinyProxy to check the health of the container, similar to the heartbeat from the user side.

**Comment by LEDfan** [comment]
Hi @Tylermarques 

Thank you for your insights, this is really useful. You are correct that when a container unexpectedly quits (e.g.by a OOM kill), ShinyProxy is unaware of this and still tries to proxy traffic to it. 
I don't believe we need to implement a full health check for this. Instead, when your browser performs an HTTP request and ShinyProxy notices that it's unable to connect to the container, we could gracefully handle this and offer the user to restart the container.  This should not be too complex to implement. In fact, we are currently doing some refactoring for the next release and this could become part of it. Hence, I'll try to make this part of the next release.

**Comment by LEDfan** [comment]
Hi, today we released ShinyProxy  3.0.0 which contains a few improvements related to this issue:

- after (re-)login users will always end up on the main page, to prevent that they end up on some internal endpoint
- ShinyProxy now notices when an app has crashed and displays a nice error
- the process to start an app has been made more stable. E.g. a user can now refresh the page while the app is starting up (previously this would cause some weird issues).
